### DISCUSSION TITLE
Blank Dashboard in New Installation

### DISCUSSION THREAD

**Discussion by FlaMike** [tagyoureit/nodejs-poolController#591] [App and O/S Config Questions] [answered]
I am brand new to the nodejs-poolController and am having what I believe are configuration issues.  I am running the system on a Pi 3B+ located in the house and communicating with a Pentair Easytouch system with Moyina USB to RS485 converter, terminal block to RJ45 adapters with Cat6 cable between the adapters.  I am getting good communication between the Pi and the Easytouch, with failure rate well below 5%.  I followed the Trouble Free Pools guide to set up the system.  

The nodejs-poolController software recognized the Easytouch controller upon startup.  The dashPanel started without errors.  However, when I browse to the web interface I get the waterdrop background image and a blank control panel (see attached screenshot).  The panel does not respond to mouse clicks and refreshing the page results in the same screen.

I am stuck and do not know where to go from here.  Can you please help me get the dashPanel web interface working?

Thank you for your help and suggestions!

![PoolControllerDashboard](https://user-images.githubusercontent.com/41341956/188331680-91be35d5-9fb8-4fc6-88be-451b2d0562f8.png)

**Comment by FlaMike** [comment]
P.S.  I am running version 7.7.0 of both the pool controller and the dash panel.

**Reply by tagyoureit** [reply]
Can you open up the developer tools and look in the Java script console and see if there are any errors?  Are you running njsPC and dP on the same box or different machines?

**Comment by FlaMike** [comment]
I have the following in the console:

![image](https://user-images.githubusercontent.com/41341956/188334726-0ffadd50-1aad-4e5f-a167-077f77b80513.png)

**Comment by FlaMike** [comment]
There are no gears on the screen, see attached.

Do I need to change the IP address in the dashpanel config.json.  I am not viewing the dashpanel from the pi?
![image](https://user-images.githubusercontent.com/41341956/188334744-1a2c5a37-e71c-459d-8096-db0ea973313d.png)

**Comment by FlaMike** [comment]
In Firefox, it looks like the Settings window appears momentarily & then disappears.  Not sure it's the settings window because it's very small & blank.  I did change the http address from 0.0.0.0 in the config.json of the dashpanel.  It did not appear to help, however.  I did restart after changing the config file.

Is there a way to tell the dashpanel the URL of the pool controller through its config.json file?  Both are on the same machine, but I can't get to the settings in the web UI of the dashpanel.  Never mind, it's already in the dashpanel config file correctly.

**Reply by tagyoureit** [reply]
The 0.0.0.0 address tells the HTTP server to accept connections from any client.  The `services.ip` should be the address of your njsPC installation.  All of this is configurable from the UI so you don't need to guess which settings are for what.

From your console screenshot above it looks like njsPC should be at `192.168.1.190`.  But there is no data being returned from `http://192.168.1.190:4200/state/all` which means that njsPC isn't at that address or is not returning data.  Can you verify the njsPC address and see if it returns data at the full url I listed?

Normally when dP can't find njsPC you will be prompted with a dialogue box like this:
![2022-09-04_17-32-54](https://user-images.githubusercontent.com/1646840/188339818-df42a462-d273-49d7-8274-2cae37979764.png)

The `find server` button can try to locate the njsPC app via ssdp, or you can type it in manually here (instead of modifying the config.json).

**Comment by FlaMike** [comment]
That is definitely the correct URL for njsPC.  Both the server and the dashboard are at that address (the Pi's URL with appropriate ports).

If I enter http://192.168.1.190:4200/state/all in a browser, there is no response--it just spins & waits for a response.

Here is the result from the Pi when I query the active processes.  The first npm start is the pool controller & the second is the dashboard.

ps -a
  PID TTY          TIME CMD
  525 tty1     00:00:00 bash
  665 pts/0    00:00:03 npm start
  676 pts/0    00:00:00 sh
  699 pts/0    00:00:05 node
  714 pts/0    00:00:20 npm start
  725 pts/0    00:00:00 sh
  747 pts/0    00:00:04 node
  876 pts/0    00:00:00 ps

I don't know why the pool controller does not respond.

I placed the two config.json files in a file posted in the next message because the inline formatting was messed up.

**Comment by FlaMike** [comment]
It looks like the formatting of the .json files was messed up.  Hopefully this is better.

[configfiles.txt](https://github.com/tagyoureit/nodejs-poolController/files/9485952/configfiles.txt)

**Reply by tagyoureit** [reply]
If you aren't getting a response on the /state/all endpoint you need to figure out what is wrong with njsPC.  You should make a copy/backup of the files in the ./data directory (poolConfig.json and poolState.json) and let the app gather all of the equipment and settings from EasyTouch again.  Check the logs from njsPC and the errors as well.

**Comment by FlaMike** [comment]
Thank you for your guidance.  I will do as you suggest & hopefully get things going before too long.

I apologize for taking up so much of your time, particularly on a holiday weekend.  You efforts in creating the pool controller and your very generous time & help in making it available & assisting nephytes like me are greatly appreciated!!!

**Reply by rstrouse** [reply]
I just got back in town.  @FlaMike you should upload the two files that are in the data directory (`poolConfig.json` and `poolState.json`) and perhaps that will shed some light on this.

Also just to be sure you are not starting njsPC then immediately stopping it by running it in the foreground from an SSH terminal then closing it or starting it in a command window from a remote desktop then closing the command window.  Both njsPC and dashPanel are server processes.

**Comment by FlaMike** [comment]
Thank you for getting back to me when you returned home.  The two files are attached as you requested, saved as text files to keep github happy.

To start both the pool controller and the dashboard, I used the following command in each program's home directory, without the quotes, of course.  "npm start&"

I accessed the pi via ssh and did not log out or exit the terminal emulator program while the programs were running.  
I also tried using a GUI version of the pi's OS and ran the software in the same manner from a local terminal window.  In that  case, I used the pi's local web browser to access the dash panel, with the same results--an empty dashpanel as shown in an earlier message above.

Finally, I tried using a docker install and got a similar result.

I also installed version 7.4 in case there  was  a difference in the current  version that caused the problem I experienced.  However, the dashboard was blank again.  (I installed fresh copies of the OS and your software for each experiment I ran.)

Originally, I was trying to use your pool controller software because the  protocol converter for my pool broke 2 months ago.  After waiting for a  warranty replacement for several weeks, I found your software & hoped it would allow me to regain control of  the pool.  Yesterday, the replacement protocol converter arrived & my pool is back to normal operation.

I plan to revisit your system, perhaps in the winter, when I can do so at a more relaxed pace.  It seems like it's a good long-term solution and I look forward to using it, once I can get it to be operational.  Any & all suggestions you have are most welcome!

Thank you again for your hard work in creating this system and your diligence and help in supporting  it.


[poolConfig.txt](https://github.com/tagyoureit/nodejs-poolController/files/9529675/poolConfig.txt)
[poolState.txt](https://github.com/tagyoureit/nodejs-poolController/files/9529676/poolState.txt)

**Reply by tagyoureit** [reply]
Nothing is jumping out to me at the moment.  But there is no reason to go back to 7.4.  That version is 1 year old and we are constantly making updates/changes.

I'd suggest you start from scratch.  Delete all the files and all the njsPC and dashpanel directories.  Pull everything new (the latest versions) and setup your adapter via the GUI.  99% of the time this app is just plug-n-play.

**Reply by rstrouse** [reply]
@tagyoureit may be on to something I was able to load your data files on a test system and got the following screen.  The only weirdness I noticed was that your spa circuit does not have a name.
![image](https://user-images.githubusercontent.com/47839015/189234942-483f64c2-9206-468d-ada7-4296606c6014.png)

**Comment by FlaMike** [comment]
Thank you both for getting back to me.  Wow!  If I had seen the screen you posted above I would have done cartwheels!  I am not surprised about the "unnamed" spa--I don't have one.  Perhaps Pentair's defaults always include a spa or the installer set up a ghost spa without thinking?  I should change that setting the Easytouch.

The other thing that caught my eye is there is no output setting for the salt cell, it should have been 28%, but that may not be displayed since there was no connection to a pool.  (Although there are temperature, equipment,  and time settings appear--snapshots from the configuration, perhaps?)

I am still scratching my head as to why the pool controller program either terminated or did not respond to the dash panel program.  Each time I hit a brick wall, I erased and loaded the SD card with a fresh copy of the OS & then followed the instructions on either your github page or the Troublefree Pool blog post to install your system.  It appeared to me that both programs completed installation and were listening on their respective ports.  The only deviation I'm aware of was the blog post did not have a step for installing git prior to pulling the program images.  Easy enough to add into the mix, however.

You also suggest setting up the adapter from the GUI.  I'm not sure I understand how to do that, as I thought the adapter had to be "online" prior to starting the pool controller program.  My adapter was located at TTYUSB0, so the default was fine as far as I can tell.

I will give things another go when I get some time to do a deep dive.  I'd really like to get your system going as it really looks like a winner!  Any other suggestions you have are most welcome.  I feel like I'm within an eyelash of getting it going.

Thank both of you so much for your help &  responses!  I truly appreciate it.

**Reply by tagyoureit** [reply]
I don't have screenshots that I can share, but the first screen you'll see in dashPanel is the one I posted above.  Once you connect to njsPC you can change the adapter options.  njsPC will load as a standalone controller (named Nixie) and be happy to setup an entire pool or do absolutely nothing.  However, once the adapter sees a known OCP it will immediately flip and be a submissive, obedient partner that will listen and support your physical OCP.

**Reply by rstrouse** [reply]
The output setting for the chlorinator and pump being blank is because my test rig does not have a chlorinator or pump attached.  It listens for changes to these outputs and if it doesn't hear any it won't pick it up.  So this is normal.  The state and configuration cache is stored in those two files you uploaded and that is where the rest of your configuration is coming from.

Once you get up and running you will be able to hide the spa body and you are correct that EasyTouch always reports a spa body.  What I have not seen is a case where the spa body didn't have a name identifier.  For at least the body circuits typically there is always a name id.  By chance do you have an EasyTouch2 4P instead of an EasyTouch 2 4 (with P at the end)?

**Comment by FlaMike** [comment]
From the command line, I saw that Nixie set up when I started the dashPanel.  Could this have been a result of the poolController disappearing/not responding?  The settings window wasn't available in my situation, as you can see from the screenshots, so that's  probably why I didn't have adapter settings available.

Still flummoxed as to why the poolController stopped or became unresponsive...  The only complaints I recall from installation and startup related to location not being set.  I know it can be done from the GUI or the config file.  I think that was more of a "warning" the impacts the system knowing when sunset, sunrise, and the like occur.

Since I have the two config files and they seem more or less OK, can I create a test rig with a pi, the programs, and adapter to help overcome my challenges or do I need pool hardware to respond via the adapter?

The controller installed is an EasyTouch 2 4, not a 2 4P.  It's firmware is  POOL: 5.2 Build 738.0 Rel

I downloaded the "diagnostics" from the EasyTouch.  Is there anything I can look for in it that may shed light on the poolController issues I'm having?

**Comment by rstrouse** [comment]
Those diagnostics are not going to do you any good since the issue simply looks like the response from njsPC is blank.  I am curious, does the following url come up `http://192.168.1.190:5150/messageManager.html`.  Instructions for this page are in the following wiki.  This should show comms from the EasyTouch OCP.
https://github.com/rstrouse/nodejs-poolController-dashPanel/wiki/Message-Manager

Yes you can create a test rig and it can be installed on a local PC or Mac if you like.
https://github.com/tagyoureit/nodejs-poolController/wiki/Quick-Start-Guide

**Reply by rstrouse** [reply]
As an aside this will also allow you to add a chem doser pump that keeps track of how much chemical has been dispensed over time.  I noticed you had a circuit name Chemical Feeder if that is what that is for.
![image](https://user-images.githubusercontent.com/47839015/189415462-6b665c44-7222-41f5-be64-be1584b02467.png)

**Comment by FlaMike** [comment]
Thanks again for your responses.   I may set  up a test rig to help get things working & then transfer to the real thing.

At the moment, I am only running the stock Pentair controller & have (at least temporarily) removed everything related to your nodejs controller/dashboard.  I didn't want the "tech" who was replacing the defective protocol converter to see any "foreign" equipment attached to the Pentair stuff, lest he conclude I broke the equipment.  So, I can't answer your question about the messagemanager.html page at this time.

The chemical feeder circuit in the current system just turns the acid pump on/off.  The pump is set manually via a crude dial to control volume per dose.  It's OK once you dial it in, but certainly not the preferred embodiment.  On demand is far superior.

My previous house had a contorller that measured ph directly & added acid as needed.  I would be delighted to re-create a similar system, perhaps with an Atlas Scientific ph probe & the pump you're referring to?  If you have the opportunity to send me more information about the acid pump, I would appreciate it greatly.

I am always amazing how crude (and out-of-date and over priced) pool equipment is.  Designed by guys using ladder logic decades ago & never updated other than transferring the old mess to newer (but obsolete) electronics.  Pool lights are an amazing example.  You *really* change colors/patterns by toggling on/off?  Oh for the option of running WLED for my lights!  But even the fanciest & newest pool LEDs & controllers toggle to change the limited colors & patterns.  Maddening!

Your nodejs system is my hope for the future!

**Reply by rstrouse** [reply]
Don't get me started on the stupidity of pool lighting.  I have my landscape lighting tied into DMX and it even does sound to light processing so when the Who is playing on the stereo the lights are dancing around the yard like you are at a concert.  When my IntelliBrites finally bite it (or I just get tired of bitchin) I am going to design my own for the replacement boards that use DMX decoders for the light channel control.

njsPC supports both types of chemistry controllers.  It will take your existing dumb pump like IntelliPh and allow you to add some smarts to it and it does on-demand dosing like IntelliChem on steriods.  The screenshot above is the time over volume based dosing strategy and the on-demand chem controller can be found here.
https://github.com/tagyoureit/nodejs-poolController/wiki/DIY-REM-Chem-Controller

**Comment by FlaMike** [comment]
Sweet yard lighting!  I have sound reactive WLED on toe kicks by the TV (and WLED based hyperion on the back of the TV).  Fun stuff for a party of 2 or 50!  I'm not trying to participate in the "best lighting control" debates.  Both systems are super and give the options of amazing fun.  I happened to start with WLED a couple of years ago & have stuck with it... at least so far.  

I run Home Assistant for most things & I know njsPC can be integrated into it.  Looking forward to that, as well.  For a variety of reasons and  history, I run Z-wave, WiFi, Zigbee, and recently a couple of BLE freezer sensors in HA.  Of  course, MQTT runs deep in that world.  HA is a super platform with lots of flexibility and  options.  I need to migrate from supervisor to Docker & have to decide if I should hold off njsPC until after HA is migrated (now that I have a working protocol converter).  Such difficult decisions :-)

**Reply by rstrouse** [reply]
Yeah DMX vs WLED is not really a debate.  You can address WLED over long run DMX or address DMX over WLED (including ArtNet).   The primary reason for DMX is the transmission distance.  It yawns at 500 feet and can go about a 1/2 mile with passive termination.

I think you will find that there are quite a few HASS users and there are lots of options for comms through MQTT, rest, and sockets if you desire.  Currently, I have about 220 Z-wave and virtual devices (mostly switches) connected to an old split Vera platform that just continues to work.  It took a bit of Lua code to make all of that talk and switching to another platform gives me hives.  

It is amazing though how much effort has been placed on a chat room protocol to make it transactional.  While MQTT is popular, it has some warts as a command interface.  I think @Crewski took a different approach and came up with a ReST/socket implementation for his HASS integration.  https://github.com/Crewski/njsPC-HA

**Comment by FlaMike** [comment, marked as answer]
Gentlemen:

I am delighted to let you know after goodness knows how many attempts, re-installations of variants of the Pi OS, nodejs-poolController, etc., replacements of RS485 converter dongles (sorry Amazon!), I *finally* got your system running!  I cannot tell what a relief it is!  I want to share with you what finally got me over the hump.

I am running a headless version of 32 bit Raspi and I ssh into the Pi 3B+.  I tried to follow to the letter the detailed instructions posted at Troublefreepool.  One minor note...  The instructions do not state that "git" needs to be installed.  That was never my stumbling point, however.  Others could probably benefit from the update (I know you did not write the instructions posted there).

The instructions state that after starting the poolController server, once the "configuration completed" output is posted, to hit return, change directories to the dashPanel directory, and do an npi i followed by another npm start& command.

Based on my experience, if you ssh into the Pi, hitting the return key once the poolController is running (per above) stops and/or kills the poolController program!  Hence, there is no data available to the dashPanel program.  However, if I ssh to the Pi through second terminal emulator, go to the dashPanel directory, and start the dashPanel program once the poolController is running, all is well with the world; your fine program runs happily and I can access the dashPanel via a web browser on a PC on the local network.  I also set up the correct time zone, longitude, and latitude.  All of the controllers circuits, programs, etc.  appear to be present as expected.

If I logout from the terminal emulator, the program running from that window terminates.  Is there a way to keep the poolConrtoller and dashPanel software running without remaining logged in?  I am using npm start& to get the client and server running.

I have a couple of other questions for you.

I would like to get MQTT going so I may integrate the poolController into Home Assistant.  My MQTT server has a username and password associated with it.  I do not see a place in UI for entering this information.  Consequently, the poolController cannot log into my MQTT server.  Can you please let me know how to enter an MQTT username and password?

Ultimately, I would like to run the poolController in a docker container.  I tried to get this approach working.  However, I got "permission denied" errors when attempting to access the USB converter.  I don't know how to overcome this challenge.  Any help and suggestions you have are most appreciated.

Thanks again for your patience & help!  I look forward to using the poolController from within Home Assistant and your software running on docker once I get the details of how to accomplish it worked out.

**Comment by FlaMike** [comment]
OK, I found the answer to one of my questions.  If I change the start command to include the no hangup command, I can log out of the pi & the programs keep running.  i.e., nohup npm run start:cached&

Progress!

**Reply by rstrouse** [reply]
The other thing to do is use pm2.  It will restart the process if it halts and on reboot of the pi.

**Comment by FlaMike** [comment]
Thanks!  Getting there bit by bit :-)

Any off the cuff help regarding MQTT username/password?  Can I place them in a config file if there's no UI available?

**Reply by Crewski** [reply]
I don’t have a quick answer for MQTT, but if you wanted to try a home assistant integration that uses socketio and restful call you can try my custom integration.  
https://github.com/Crewski/njsPC-HA

**Reply by tagyoureit** [reply]
> Any off the cuff help regarding MQTT username/password? Can I place them in a config file if there's no UI available?

Stop njsPC.  Edit `./config.json`.  Find the `mqtt` section and under options add the username/password.  Restart njsPC.

![image](https://user-images.githubusercontent.com/1646840/192126764-224d48f9-7c13-4588-87aa-f6e77d67d7ac.png)

**Comment by FlaMike** [comment]
Thank you!

**Comment by FlaMike** [comment]
Got pm2 going.  Thank you.   Now onto Crewski's HA integration...

**Comment by FlaMike** [comment]
Forgive me for being naive.  It's been a while since I've installed a custom component in HACS.  I copied and added the repository & the files into the custom components directory of HACS.  I've restarted Home Assistant.  When I go to HACS Integrations, njsPC-HA shows up as a new repository but I cannot configure it.  If I click on it, I get to the instructions page for the component.  There is no install option.  I'm running version 1.27.2 of HACS and 2022.8.7 of Home Assistant.

**Reply by Crewski** [reply]
No worries.  If you manually copied it into custom_components you might want to delete it out.  I'll run though the steps "from scratch".

1. In HACS go to Integrations.  Click on the 3 dots in the top right and go to Custom Repositories
![image](https://user-images.githubusercontent.com/8012153/192121905-a4efb9aa-06ec-44e0-89bc-755d49664ebd.png)

2. Add `https://github.com/Crewski/njsPC-HA` into the Repository field, select integration, then click Add.
![image](https://user-images.githubusercontent.com/8012153/192121928-ecafe3e1-c3cd-46b0-8cda-a6bacba8bd9a.png)

3. I don't remember if that will install it automatically or if you need to click Explore and Download Repositories in HACS.
4. Restart Home Assistant
5. Go to Integrations, select Add Integration in the bottom right, and then search for njsPC-HA.  If it doesn't show up, refresh your browser with CTRL+F5.  Hopefully it'll show up then.

**Comment by FlaMike** [comment]
Thank you.  I will start from scratch & hope to report success to you
tomorrow!

**Comment by FlaMike** [comment]
I deleted the existing njspc folder & its contents in custom components, deleted the custom repository from yesterday's attempts & then started from scratch as you suggested.  FYI, I did have to hit "Explore & Download Repositories" after adding the github URL.  I restarted HA, got the integration & now have 20 entities in HA that I can use to monitor and control the pool equipment (haven't added them to the UI yet, but I assume all will be fine now).

Thank you so much for your suggestions, help, & patience!!!!

As an aside, I will add the MQTT login credentials & see if I can get them going as well.

Good start to the day!!!

**Comment by FlaMike** [comment]
Just stopped the pool controller, edited the config.json file to add MQTT username & password, & restarted the pool controller.  Looks like all is well.there now, too.

Thanks again to all.  Very happy to be using your system!  Much more to play with, but I feel like I'm on my way :-)

**Comment by keithpre** [comment]
I'm really late to the party, but since I also ran into this, I thought I'd leave a note to the next person.   I was seeing this exact same symptom after the IP address of my controller changed.   The UI was non-functional and wasn't able to allow me to update the IP address of the controller.    I ended up manually needing to update the config for the IP address through the config.json.   Everything worked after that.
